java-frontend/src/main/java/org/sonar/java/se/AlwaysTrueOrFalseExpressionCollector.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+
+import org.sonar.java.se.constraint.BooleanConstraint;
+import org.sonar.java.se.constraint.Constraint;
+import org.sonar.java.se.constraint.ObjectConstraint;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.Collection;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+class AlwaysTrueOrFalseExpressionCollector {
+
+  private Deque<EvaluatedConditions> evaluatedConditions = new LinkedList<>();
+
+  public void init() {
+    evaluatedConditions.push(new EvaluatedConditions());
+  }
+
+  public void evaluatedToFalse(Tree condition, ExplodedGraph.Node node) {
+    evaluatedConditions.peek().evaluatedToFalse(condition, node);
+  }
+
+  public void evaluatedToTrue(Tree condition, ExplodedGraph.Node node) {
+    evaluatedConditions.peek().evaluatedToTrue(condition, node);
+  }
+
+  public void interruptedExecution() {
+    evaluatedConditions.pop();
+  }
+
+  public CheckerContext.AlwaysTrueOrFalseExpressions alwaysTrueOrFalseExpressions() {
+    return evaluatedConditions.peek();
+  }
+
+  private static class EvaluatedConditions implements CheckerContext.AlwaysTrueOrFalseExpressions {
+    final Multimap<Tree, ExplodedGraph.Node> evaluatedToFalse = HashMultimap.create();
+    final Multimap<Tree, ExplodedGraph.Node> evaluatedToTrue = HashMultimap.create();
+
+    void evaluatedToFalse(Tree condition, ExplodedGraph.Node node) {
+      evaluatedToFalse.put(condition, node);
+    }
+
+    void evaluatedToTrue(Tree condition, ExplodedGraph.Node node) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER') Rename method "evaluatedToTrue" to prevent any misunderstanding/clash with field "evaluatedToTrue" defined on line 66. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1845)